r-package/grf/R/rank_average_treatment.R

```diff
@@ -0,0 +1,297 @@
+                "(or some clusters contains too few units)."))
+  }
+
+  # For all supported forest types, DR.scores is a subset-length vector
+  # TODO future support for multi_arm_causal_forest could potentially restrict to a single (outcome, contrast).
+  DR.scores <- get_scores(forest, subset = subset, debiasing.weights = debiasing.weights,
+                          compliance.score = compliance.score, num.trees.for.weights = num.trees.for.weights)
+
+  # *** Compute the TOC and RATE ***
+
+  if (target == "AUTOC") {
+    wtd.mean <- function(x, w) sum(x * w) / sum(w)
+  } else if (target == "QINI") {
+    wtd.mean <- function(x, w) sum(cumsum(w) / sum(w) * w * x) / sum(w)
+  }
+
+  # Compute estimates, a function to be passed on to boostrap routine.
+  # @data: a data.frame with the original data, column 1: DR.scores*sample.weights, column 2: sample.weights,
+  #   column 3: priority scores (integer vector)
+  # @indices: a vector of indices which define the bootstrap sample.
+  # @returns: an estimate of RATE, together with the TOC curve.
+  estimate <- function(data, indices, q) {
+    # let q be a fraction in (0, 1].
+    # we have 1) TOC(q; Sj) = 1/[qn] sum_{i=1}^{[qn]} Gamma_{i(j)} - ATE
+    # and 2) RATE = 1/n sum_{i=1}^{n} TOC(i/n; Sj)
+    # For boostrapping the TOC curve, we fix q on some grid q'.
+    # For estimating the RATE we set 1/n <= q <= 1.
+    # Observations:
+    # a) the entire TOC curve can be computed as a cumulative sum of sorted DR.scores.
+    # b) taking ties into account amounts to using the average DR.scores within each
+    # tied group instead of the individual DR.scores.
+    # So the steps are:
+    # Compute average DR.scores by priority group in increasing order: DR.scores',
+    # repeat the possibly tied entries with the number of duplicates.
+    # Take the cumsum of this divided by ni to get the TOC curve. Take a (weighted) average
+    # of this to get RATE. This is what is being done below, using R's fastest primitives for
+    # quick aggregation and grouping (using for example "rowsum"),
+    nq <- floor(length(indices) * q)
+    grid.id <- rep.int(nq, c(nq[1], diff(nq)))
+
+    prio <- data[indices, 3]
+    idx <- order(prio, decreasing = TRUE)
+    sample.weights <- data[indices, 2][idx]
+
+    num.ties <- tabulate(prio) # count by rank in increasing order
+    num.ties <- num.ties[num.ties != 0] # ignore potential ranks not present in BS sample
+    grp.sum <- rowsum(data[indices, 1:2][idx, ], prio[idx], reorder = FALSE)
+    DR.avg <- grp.sum[, 1] / grp.sum[, 2]
+
+    DR.scores.sorted <- rep.int(DR.avg, rev(num.ties))
+    DR.scores.grid <- rowsum(cbind(DR.scores.sorted * sample.weights, sample.weights), grid.id, reorder = FALSE)
+    ATE <- sum(DR.scores.sorted * sample.weights) / sum(sample.weights)
+
+    TOC <- cumsum(DR.scores.sorted * sample.weights) / cumsum(sample.weights) - ATE
+    TOC.grid <- cumsum(DR.scores.grid[, 1]) / cumsum(DR.scores.grid[, 2]) - ATE
```

Comment:
So this line is potentially problematic. I'll follow up via email as well to discuss the math, but basically with sample weights it gets tricky to think about TOC(_u_) for some arbitrary _u_ because when you randomly permute the samples that have tied priorities, if _u_ falls within that set of ties for the weighted empirical CDF, you may get a situation where under some permutations _k_ of the tied samples fit under _u_ but under others only _k_' fit.

Reply:
Got it, the grid calculations have been revised to allow for arbitrary query points q using fractional samples